src/details/ArborX_DetailsHalfTraversal.hpp

```diff
@@ -0,0 +1,88 @@
+/****************************************************************************
+ * Copyright (c) 2017-2022 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_DETAILS_HALF_TRAVERSAL_HPP
+#define ARBORX_DETAILS_HALF_TRAVERSAL_HPP
+
+#include <ArborX_DetailsHappyTreeFriends.hpp>
+#include <ArborX_DetailsNode.hpp> // ROPE_SENTINEL
+
+namespace ArborX::Details
+{
+
+template <class BVH, class PredicateGetter, class Callback>
+struct HalfTraversal
+{
+  BVH _bvh;
+  PredicateGetter _predicate;
+  Callback _callback;
+
+  template <class ExecutionSpace>
+  HalfTraversal(ExecutionSpace const &space, BVH const &bvh,
+                PredicateGetter const &predicate, Callback const &callback)
```

Comment:
Would we also want a default here for `get_predicate` at some point?

Reply:
Made it trailing argument